File: javascript/programmers/javascript/level2/괄호변환.js
const p = "()))((()";

const isCorrectString = (p) => {
  let check = 0;
  for (let i = 0; i < p.length; i++) {
    if (p[i] === "(") {
      check += 1;
    } else {
      check -= 1;
    }
    if (check < 0) {
      return false;
    }
  }
  return true;
};

const findBalanceStringNum = (p) => {
  let check = 0;
  for (let i = 0; i < p.length; i++) {
    if (p[i] === "(") {
      check += 1;
    } else {
      check -= 1;
    }
    if (check === 0) {
      return i;
    }
  }
};

const reverseString = (p) => {
  let answer = "";
  for (let i = 0; i < p.length; i++) {
    if (p[i] === "(") {
      answer += ")";
    } else {
      answer += "(";
    }
  }
  return answer;
};

const Recursive = (p) => {
  if (p === "") {
    return "";
  }
  if (isCorrectString(p)) {
    return p;
  } else {
    const idx = findBalanceStringNum(p);
    let u = p.substring(0, idx + 1);
    let v = p.substring(idx + 1);
    if (isCorrectString(u)) {
      return u + Recursive(v);
    } else {
      let temp = "(" + Recursive(v) + ")";
      let temp2 = reverseString(u.substring(1, u.length - 1));
      return temp + temp2;
    }
  }
};

const solution = (p) => {
  let answer = "";
  answer = Recursive(p);
  return answer;
};

console.log(solution(p));

File: python/programmers/level2/괄호 변환.py
def isCorrect(s):
    num = 0
    for x in s:
        if x == '(':
            num += 1
        else:
            num -= 1
        if num < 0:
            return False
    if num == 0:
        return True
    else:
        return False


def logic(p):
    # 1. 입력이 빈 문자열인 경우, 빈 문자열을 반환
    if len(p) == 0:
        return ""
    elif isCorrect(p):
        return p
    # 2. w를 두 균형잡힌 괄호 문자열 u,v 로 분리한다.
    num = 0
    idx = 0
    while True:
        if p[idx] == '(':
            num += 1
        else:
            num += -1
        if num == 0:
            break
        idx += 1
    u = p[0:idx + 1]
    v = p[idx + 1:]
    # print(u, "그리고", v)
    if isCorrect(u):
        u += logic(v)
        return u
    else:
        tmp = "("
        tmp += logic(v) + ")"
        tmp += ''.join(list(map(lambda x: '(' if x == ')' else ')', u[1:-1])))
        return tmp


def solution(p):
    answer = logic(p)
    return answer


print(solution("()))((()"))
